Library/Homebrew/keg.rb

```diff
@@ -647,6 +647,15 @@ def make_relative_symlink(dst, src, **options)
     dst.delete if options[:overwrite] && (dst.exist? || dst.symlink?)
     dst.make_relative_symlink(src)
   rescue Errno::EEXIST => e
+    # Retry if we're linking a different version of the same
+    # formula. The `AlreadyLinkedError` above won't catch
+    # this if a formula is missing an optlink. In that case,
+    # delete the symlink and retry.
+    if dst.symlink? && dst.exist? && Keg.for(dst).name == name
```

Comment:
Unfortunately `Keg.for(dst).name` can raise an exception which can prevent `brew link --overwrite` from overwriting anything without a keg.